src/psyclone/dynamo0p3.py

```diff
@@ -5159,38 +5159,93 @@ def _compute_halo_depth(self):
+        ignored.
+
+        :param bool ignore_hex_dep: if True then ignore any read \
+        accesses contained in halo exchanges. This is an optional \
+        arguments that defaults to False.
 
         :return: a list containing halo depth information derived from \
         all fields dependent on this halo exchange
         :rtype: :func:`list` of :py:class:`psyclone.dynamo0p3.HaloDepth`
 
         '''
         # get our halo information
-        halo_info_list = self._compute_halo_read_info()
+        halo_info_list = self._compute_halo_read_info(ignore_hex_dep)
         # use the halo information to generate depth information
         depth_info_list = _create_depth_list(halo_info_list)
         return depth_info_list
 
-    def _compute_halo_read_info(self):
+    def _compute_halo_read_info(self, ignore_hex_dep=False):
         '''Dynamically computes all halo read dependencies and returns the
-        required halo information (i.e. halo depth and stencil type) in a
-        list of HaloReadAccess objects
+        required halo information (i.e. halo depth and stencil type)
+        in a list of HaloReadAccess objects. If the optional
+        ignore_hex_dep argument is set to True then any read accesses
+        contained in halo exchange nodes are ignored.
+
+        :param bool ignore_hex_dep: if True then ignore any read \
+        accesses contained in halo exchanges. This is an optional \
+        arguments that defaults to False.
 
         :return: a list containing halo information for each read dependency
         :rtype: :func:`list` of :py:class:`psyclone.dynamo0p3.HaloReadAccess`
 
+        :raises GenerationError: if there is more than one read \
+        dependency associated with a halo exchange.
+        :raises GenerationError: if there is a read dependency \
+        associated with a halo exchange and it is not the last entry \
+        in the read dependency list.
+
+        :raises GenerationError: if there is a read dependency \
+        associated with an asynchronous halo exchange.
+
+        :raises GenerationError: if no read dependencies are found.
+
         '''
         read_dependencies = self.field.forward_read_dependencies()
+        hex_deps = [dep for dep in read_dependencies
+                    if isinstance(dep.call, DynHaloExchange)]
+        if ignore_hex_dep and hex_deps:
+            # There is a field accessed by a halo exchange that is
+            # a read dependence. As ignore_hex_dep is True this should
+            # be ignored so this is removed from the list.
+            if any(dep for dep in hex_deps
+                   if isinstance(dep.call, (DynHaloExchangeStart,
+                                            DynHaloExchangeEnd))):
+                raise GenerationError(
+                    "Please perform redundant computation transformations "
+                    "before asynchronous halo exchange transformations.")
+
+            # There can only be one field accessed by a
```

Comment:
Yes, a halo exchange should fully cover a dependency otherwise you would need more than one halo exchange for the same dependency. If your halo was dirty to level 2, you would only want to perform a single halo exchange to make the halo clean. Therefore there should only ever be one.